cmd/gcp-controller-manager/app/gke_approver.go

```diff
@@ -189,6 +200,10 @@ type csrValidator struct {
 	validate validateFunc
 
 	permission authorization.ResourceAttributes
+
+	// precondition is an optional field that can perform operations and verify an approval should continue.
+	// If precondition returns an error, CSR will be retried.
+	precondition preconditionFunc
```

Comment:
Can we re-use `validate` instead?

Reply:
we need additional information (the client), and I wanted the precondition to run last, since it is the most expensive

Reply:
You can add more arguments to `validate` and put these checks at the end of validation.
That only leaves SAR check after this validation.
I'd like to keep the abstraction here simple, especially since this only affects a single type of CSRs.

Reply:
> That only leaves SAR check after this validation.

taking a disruptive action prior to verifying the CSR submitted was authorized doesn't seem appropriate

Reply:
we also want metrics specifically around this check, and the validate check doesn't log metrics in error/retry cases

Reply:
Ugh, ok. Let's at least change the naming. `precondition` sounds like something that runs before recognize/validate.
Maybe `validateAfterSAR`?

Reply:
change to `preApproveHook` and updated doc